src/reducers/computeRenderedRows.js

```diff
@@ -239,3 +245,97 @@ function computeRenderedRowOffsets(state, rowRange) {
   state.rowOffsets = rowOffsetsCache;
   state.rows = bufferMapping;
 }
+
+
+/**
+ * Walk the rows to render and compute the height offsets and
+ * positions in the row buffer.
+ *
+ * NOTE (jordan) This alters state so it shouldn't be called
+ * without state having been cloned first.
+ *
+ * @param {!Object} state
+ * @param {{
+ *   endBufferIdx: number,
+ *   endViewportIdx: number,
+ *   firstBufferIdx: number,
+ *   firstViewportIdx: number,
+ * }} rowRange
+ * @private
+ */
+function computeRenderedRowOffsetsInViewport(state, rowRange) {
+  const { rowBufferSet, rowOffsetIntervalTree, storedHeights } = state;
+  const {
+    endBufferIdx,
+    endViewportIdx,
+    firstBufferIdx,
+    firstViewportIdx,
+  } = rowRange;
+
+  const renderedRowsCount = endBufferIdx - firstBufferIdx;
+  if (renderedRowsCount === 0) {
+    state.rowOffsets = {};
+    state.rows = [];
+    return;
+  }
+
+  // output for this function
+  const rows = state.rows.slice(); // state.rows
+  const rowOffsets = {}; // state.rowOffsets
```

Comment:
I think this still permits the rows & rowOffsets to have different lengths / members the way we're tackling this.  Could we modify to keep them in sync or otherwise improve it to avoid any bugs related from assuming they're in sync?

Reply:
This is true. I'll change it.

Reply:
This whole concern may go away if we address things the way I describe in my other comment above.

Reply:
Now they only have indexes and offsets for the rows inside the viewport.